### PR TITLE
ci: Improve dynamic machine setup especially for randomized AD

### PR DESCRIPTION
**Pull request by jakub-vavra-cz**
Improve dns role to be less reliant on idmci and work with machines more dynamically.
Update ipa and client roles to handle better missing backends.
Change ad role to allow hostname/domain randomization. 
Remove partially service.XXX style config in a favour of dynamic configuration. 
Refactor joining domains (IPA, AD samba) to standalone files.

### CODE REVIEW

**Comment by pbrezina** [comment]
LGTM, but I see the following error when running `vagrant up`:

```
TASK [ad : Create new AD forest ad.test] ***************************************
fatal: [dc.ad.test]: FAILED! => changed=true 
  cmd: |-
    Import-Module ADDSDeployment
  
    Install-ADDSForest                                                         -DomainName "ad.test"                                    -CreateDnsDelegation:$false                                              -DomainNetbiosName "DC"                            -ForestMode "WinThreshold"                                               -DomainMode "WinThreshold"                                               -Force:$true                                                             -InstallDns:$true                                                        -NoRebootOnCompletion:$true                                              -SafeModeAdministratorPassword                                             (ConvertTo-SecureString 'vagrant' -AsPlainText -Force)
  delta: '0:00:11.031453'
  end: '2023-12-07 13:54:15.505101'
  msg: non-zero return code
  rc: 1
  start: '2023-12-07 13:54:04.473648'
  stderr: |-
    Install-ADDSForest : Verification of prerequisites for Domain Controller promotion failed. The NetBIOS name DC is
    already in use.
    At line:3 char:1
    + Install-ADDSForest                                                    ...
    + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : NotSpecified: (:) [Install-ADDSForest], TestFailedException
        + FullyQualifiedErrorId : Test.VerifyDcPromoCore.DCPromo.General.32,Microsoft.DirectoryServices.Deployment.PowerSh
       ell.Commands.InstallADDSForestCommand
  stderr_lines: <omitted>
  stdout: |2-
  
    Message                                                                                                        Context
    -------                                                                                                        -------
    Verification of prerequisites for Domain Controller promotion failed. The NetBIOS name DC is already in use... Test....
  stdout_lines: <omitted>
```

The error does not happen if I run vagrant up from master branch.

**Comment by pbrezina** [comment]
@danlavu Please, give it a try.

**Comment by jakub-vavra-cz** [comment]
Pushed change removing the line mentioning the netbios. 
Same parameter is commented out in idmci, let me know it fit fixes the issue.

File: src/ansible/roles/ipa/tasks/main.yml [hunk 1]
```diff
@@ -90,41 +101,77 @@
...
-    stdin: |
-      {{ service.ad.safe_password }}
-      {{ service.ad.safe_password }}
+- name: 'Setup trust with AD'
+  block:
+  - name: Set AD facts (IPA)
+    set_fact:
+      ad_domain: "{{ '.'.join(hostvars[groups.ad.0].inventory_hostname.split('.')[1:]) }}"
+      safe_password: "{{ service.ad.safe_password }}"
+      # TODO: range-type can be configurable ipa-ad-trust or ipa-ad-trust-posix
```
Comment (ikerexxe) [inline src/ansible/roles/ipa/tasks/main.yml:147]: Did you leave this TODO on purpose?

**Review by ikerexxe — COMMENTED** [review]
One question and I'm ready to approve

File: src/ansible/roles/ipa/tasks/main.yml (hunk 1 above)
Comment (jakub-vavra-cz) [reply src/ansible/roles/ipa/tasks/main.yml:147]: Yes, I added it on purpose I wanted to discuss if it makes a sense to make it configurable from outside and discuss the default value as well.

Comment (ikerexxe) [reply src/ansible/roles/ipa/tasks/main.yml:147]: I don't have any expertise in this area. @pbrezina any opinion?

Comment (ikerexxe) [reply src/ansible/roles/ipa/tasks/main.yml:147]: In that case, let's resolve it and we can discuss it in the future

**Review by ikerexxe — APPROVED** [review]
LGTM!

**Pull request merged**